### PR TITLE
Problem: context must be used to create sockets

### PR DESCRIPTION
**Pull request by somdoron**
Suggested solution: global context, now sockets can be created without a context.

### CODE REVIEW

**Comment by dalebrubaker** [comment]
Seems too good to be true. :) Is there still a purpose for context? (In other words, what does this really mean?) Thanks.

File: src/NetMQ/Global.cs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            }
+        }
+
+        internal static SocketBase CreateSocket(ZmqSocketType socketType)
+        {
+            lock (s_sync)
+            {
+                if (s_ctx == null)
+                {
+                    s_ctx = new Ctx();
```
Comment (dalebrubaker) [inline src/NetMQ/Global.cs]: So essentially this is a singleton, and this "guarantees" there is only one context per process, but it's handled behind the scene. No need for explicit NetMQContext unless more than one per process was needed for some reason. Do I understand correctly? Thanks.

**Comment by somdoron** [comment]
Context are old concepts from zeromq, actually there is almost no use for them in NetMQ.

One problem is that now when you dispose a socket it doesn't really get disposed immediately, it queue for termination on a dedicate thread. So the only way to make sure all sockets are disposed is to call Context.Dispose. I'm going to change it so when you call Socket.Dispose it will block until the socket is disposed. However this can break some application so I need to find a way to do it without breaking anything.

File: src/NetMQ/Global.cs (hunk 1 above)
Comment (somdoron) [inline src/NetMQ/Global.cs]: Yes.

**Comment by drewnoakes** [comment]
Looks good Doron. Is this ready for merging? Would it be worth marking `NetMQContext` as obsolete too?

**Comment by somdoron** [comment]
Yes it is ready. Tomorrow I will work on make the socket dispose itself completely, currently the last disposed socket is terminate the context, however this can take time, I prefer that each socket will get disposed completely when dispose is called . Regarding making the NetMQContext obsolete, let's make this stable first and then make it obsolete.

**Pull request merged**